File: app/src/main/java/com/robifr/ledger/ui/editqueue/viewmodel/EditQueueViewModel.java
/**
 * Copyright (c) 2024 Robi
 *
 * Ledger is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Ledger is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Ledger. If not, see <https://www.gnu.org/licenses/>.
 */

package com.robifr.ledger.ui.editqueue.viewmodel;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import com.robifr.ledger.R;
import com.robifr.ledger.data.model.CustomerModel;
import com.robifr.ledger.data.model.QueueModel;
import com.robifr.ledger.repository.CustomerRepository;
import com.robifr.ledger.repository.ProductRepository;
import com.robifr.ledger.repository.QueueRepository;
import com.robifr.ledger.ui.LiveDataEvent;
import com.robifr.ledger.ui.StringResources;
import com.robifr.ledger.ui.createqueue.viewmodel.CreateQueueViewModel;
import java.util.Collections;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;
import java.util.concurrent.ExecutionException;

public class EditQueueViewModel extends CreateQueueViewModel {
  @NonNull
  private final MutableLiveData<LiveDataEvent<Long>> _editedQueueId = new MutableLiveData<>();

  @Nullable private QueueModel _initialQueueToEdit = null;

  public EditQueueViewModel(
      @NonNull QueueRepository queueRepository,
      @NonNull CustomerRepository customerRepository,
      @NonNull ProductRepository productRepository) {
    super(queueRepository, customerRepository, productRepository);
  }

  @Override
  @NonNull
  public QueueModel inputtedQueue() {
    final Long id =
        this._initialQueueToEdit != null && this._initialQueueToEdit.id() != null
            ? this._initialQueueToEdit.id()
            : null;
    return QueueModel.toBuilder(super.inputtedQueue()).setId(id).build();
  }

  @Override
  public void onSave() {
    if (this._inputtedProductOrders.size() == 0) {
      this._snackbarMessage.setValue(
          new LiveDataEvent<>(
              new StringResources.Strings(
                  R.string.text_please_to_include_at_least_one_product_order)));
      return;
    }

    this._updateQueue(this.inputtedQueue());
  }

  public void setInitialQueueToEdit(@NonNull QueueModel queue) {
    this._initialQueueToEdit = Objects.requireNonNull(queue);
  }

  @NonNull
  public LiveData<LiveDataEvent<Long>> editedQueueId() {
    return this._editedQueueId;
  }

  @Nullable
  public QueueModel selectQueueById(@Nullable Long queueId) {
    final StringResources notFoundRes =
        new StringResources.Strings(R.string.text_error_failed_to_find_related_queue);
    QueueModel queue = null;

    try {
      queue = this._queueRepository.selectById(queueId).get();
      if (queue == null) this._snackbarMessage.setValue(new LiveDataEvent<>(notFoundRes));

    } catch (ExecutionException | InterruptedException e) {
      this._snackbarMessage.setValue(new LiveDataEvent<>(notFoundRes));
    }

    return queue;
  }

  @Override
  protected void _onUpdateAllowedPaymentMethod() {
    final QueueModel inputtedQueue = this.inputtedQueue();
    final HashSet<QueueModel.PaymentMethod> allowedPaymentMethods =
        this._allowedPaymentMethods.getValue() != null
            ? new HashSet<>(this._allowedPaymentMethods.getValue())
            : new HashSet<>(Set.of(QueueModel.PaymentMethod.CASH));

    if ((inputtedQueue.status() == QueueModel.Status.COMPLETED
        && inputtedQueue.customer() != null
        && inputtedQueue.customer().isBalanceSufficient(this._initialQueueToEdit, inputtedQueue))) {
      allowedPaymentMethods.add(QueueModel.PaymentMethod.ACCOUNT_BALANCE);
    } else {
      allowedPaymentMethods.remove(QueueModel.PaymentMethod.ACCOUNT_BALANCE);
    }

    this._allowedPaymentMethods.setValue(Collections.unmodifiableSet(allowedPaymentMethods));

    // Change payment method to cash when current selected one marked as not allowed.
    if (!allowedPaymentMethods.contains(inputtedQueue.paymentMethod())) {
      this.onPaymentMethodChanged(QueueModel.PaymentMethod.CASH);
    }
  }

  @Override
  protected void _onUpdateTemporalInputtedCustomer() {
    Objects.requireNonNull(this._initialQueueToEdit);

    final QueueModel inputtedQueue = this.inputtedQueue();
    final CustomerModel customer =
        inputtedQueue.customer() != null
            ? CustomerModel.toBuilder(inputtedQueue.customer())
                .setBalance(
                    inputtedQueue
                        .customer()
                        .balanceOnUpdatedPayment(this._initialQueueToEdit, inputtedQueue))
                .setDebt(
                    inputtedQueue
                        .customer()
                        .debtOnUpdatedPayment(this._initialQueueToEdit, inputtedQueue))
                .build()
            : null;

    this._temporalInputtedCustomer.setValue(customer);
  }

  private void _updateQueue(@NonNull QueueModel queue) {
    Objects.requireNonNull(queue);

    this._queueRepository
        .update(queue)
        .thenAcceptAsync(
            effected -> {
              if (effected > 0) this._editedQueueId.postValue(new LiveDataEvent<>(queue.id()));

              final StringResources stringRes =
                  effected > 0
                      ? new StringResources.Plurals(
                          R.plurals.args_updated_x_queue, effected, effected)
                      : new StringResources.Strings(R.string.text_error_failed_to_update_queue);
              this._snackbarMessage.postValue(new LiveDataEvent<>(stringRes));
            });
  }

  public static class Factory implements ViewModelProvider.Factory {
    @NonNull private final Context _context;

    public Factory(@NonNull Context context) {
      Objects.requireNonNull(context);

      this._context = context.getApplicationContext();
    }

    @Override
    @NonNull
    public <T extends ViewModel> T create(@NonNull Class<T> cls) {
      Objects.requireNonNull(cls);

      final EditQueueViewModel viewModel =
          new EditQueueViewModel(
              QueueRepository.instance(this._context),
              CustomerRepository.instance(this._context),
              ProductRepository.instance(this._context));
      return Objects.requireNonNull(cls.cast(viewModel));
    }
  }
}

File: app/src/main/java/com/robifr/ledger/ui/editqueue/viewmodel/EditQueueViewModel.kt
/**
 * Copyright 2024 Robi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.robifr.ledger.ui.editqueue.viewmodel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.viewModelScope
import com.robifr.ledger.R
import com.robifr.ledger.data.model.QueueModel
import com.robifr.ledger.di.IoDispatcher
import com.robifr.ledger.repository.CustomerRepository
import com.robifr.ledger.repository.ProductRepository
import com.robifr.ledger.repository.QueueRepository
import com.robifr.ledger.ui.PluralResource
import com.robifr.ledger.ui.StringResource
import com.robifr.ledger.ui.UiEvent
import com.robifr.ledger.ui.createqueue.viewmodel.CreateQueueViewModel
import com.robifr.ledger.ui.editqueue.EditQueueFragment
import com.robifr.ledger.util.updateEvent
import dagger.hilt.android.lifecycle.HiltViewModel
import java.time.ZoneId
import javax.inject.Inject
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

@HiltViewModel
class EditQueueViewModel
@Inject
constructor(
    @IoDispatcher dispatcher: CoroutineDispatcher,
    queueRepository: QueueRepository,
    customerRepository: CustomerRepository,
    productRepository: ProductRepository,
    private val _savedStateHandle: SavedStateHandle
) : CreateQueueViewModel(dispatcher, queueRepository, customerRepository, productRepository) {
  private lateinit var _initialQueueToEdit: QueueModel

  private val _editResultEvent: MutableLiveData<UiEvent<EditQueueResultState>> = MutableLiveData()
  val editResultEvent: LiveData<UiEvent<EditQueueResultState>>
    get() = _editResultEvent

  init {
    // Setting up initial values inside a fragment is painful. See commit d5604599.
    _loadQueueToEdit()
  }

  override fun onSave() {
    val inputtedQueue: QueueModel = parseInputtedQueue()
    if (inputtedQueue.productOrders.isEmpty()) {
      _onSnackbarShown(StringResource(R.string.createQueue_includeOneProductOrderError))
      return
    }
    viewModelScope.launch(_dispatcher) { _updateQueue(inputtedQueue) }
  }

  override fun onBackPressed() {
    if (_initialQueueToEdit != parseInputtedQueue()) _onUnsavedChangesDialogShown()
    else _onFragmentFinished()
  }

  override fun parseInputtedQueue(): QueueModel =
      if (::_initialQueueToEdit.isInitialized) {
        super.parseInputtedQueue().copy(id = _initialQueueToEdit.id)
      } else {
        super.parseInputtedQueue()
      }

  override fun _onUpdateAllowedPaymentMethods() {
    if (!::_initialQueueToEdit.isInitialized) return super._onUpdateAllowedPaymentMethods()

    val inputtedQueue: QueueModel = parseInputtedQueue()
    val isBalanceSufficient: Boolean =
        inputtedQueue.status == QueueModel.Status.COMPLETED &&
            _uiState.safeValue.customer?.isBalanceSufficient(_initialQueueToEdit, inputtedQueue)
                ?: false
    val isTemporalBalancePositive: Boolean =
        _uiState.safeValue.customer?.let {
          // Compare with the account balance payment option
          // as if the user does it before they actually do.
          it.balanceOnUpdatedPayment(
              _initialQueueToEdit,
              inputtedQueue.copy(paymentMethod = QueueModel.PaymentMethod.ACCOUNT_BALANCE)) >= 0L
        } ?: false
    val allowedPaymentMethods: MutableSet<QueueModel.PaymentMethod> =
        _uiState.safeValue.allowedPaymentMethods.toMutableSet().apply {
          if (isBalanceSufficient && isTemporalBalancePositive) {
            add(QueueModel.PaymentMethod.ACCOUNT_BALANCE)
          } else {
            remove(QueueModel.PaymentMethod.ACCOUNT_BALANCE)
          }
        }
    _uiState.setValue(_uiState.safeValue.copy(allowedPaymentMethods = allowedPaymentMethods))
    // Change payment method to cash when current selected one marked as not allowed.
    if (!allowedPaymentMethods.contains(inputtedQueue.paymentMethod)) {
      onPaymentMethodChanged(QueueModel.PaymentMethod.CASH)
    }
  }

  override fun _onUpdateTemporalCustomer() {
    if (!::_initialQueueToEdit.isInitialized) return _onUpdateTemporalCustomer()
    val inputtedQueue: QueueModel = parseInputtedQueue()
    _uiState.setValue(
        _uiState.safeValue.copy(
            temporalCustomer =
                _uiState.safeValue.customer?.let {
                  it.copy(
                      balance = it.balanceOnUpdatedPayment(_initialQueueToEdit, inputtedQueue),
                      debt = it.debtOnUpdatedPayment(_initialQueueToEdit, inputtedQueue))
                }))
  }

  private suspend fun _selectQueueById(queueId: Long?): QueueModel? =
      _queueRepository.selectById(queueId).also {
        if (it == null) _onSnackbarShown(StringResource(R.string.createQueue_fetchQueueError))
      }

  private suspend fun _updateQueue(queue: QueueModel) {
    _queueRepository.update(queue).let { effected ->
      _onSnackbarShown(
          if (effected > 0) {
            PluralResource(R.plurals.createQueue_updated_n_queue, effected, effected)
          } else {
            StringResource(R.string.createQueue_updateQueueError)
          })
      if (effected > 0) {
        _editResultEvent.updateEvent(
            data = EditQueueResultState(queue.id), onSet = { it }, onReset = { null })
      }
    }
  }

  private fun _loadQueueToEdit() {
    viewModelScope.launch(_dispatcher) {
      // The initial queue ID shouldn't be null when editing data.
      _selectQueueById(
              _savedStateHandle.get<Long>(
                  EditQueueFragment.Arguments.INITIAL_QUEUE_ID_TO_EDIT_LONG.key())!!)
          ?.let {
            withContext(Dispatchers.Main) {
              _initialQueueToEdit = it
              onCustomerChanged(it.customer)
              onDateChanged(it.date.atZone(ZoneId.systemDefault()))
              onStatusChanged(it.status)
              onPaymentMethodChanged(it.paymentMethod)
              onProductOrdersChanged(it.productOrders)
            }
          }
    }
  }
}
